internal/rule/rule.go

```diff
@@ -0,0 +1,80 @@
+package rule
+
+import (
+	"github.com/aws/amazon-cloudwatch-agent-test/filesystem"
+	"log"
+)
+
+type Rule[T any] struct {
+	Conditions []ICondition[T]
+}
+
+func (r *Rule[T]) Evaluate(target T) (bool, error) {
+	for _, c := range r.Conditions {
+		success, err := c.Evaluate(target)
+		if err != nil {
+			log.Printf("Error was %v", err)
+			return false, err
+		}
+		log.Printf("Evaluate result: %v", success)
+		if !success {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+type ICondition[T any] interface {
+	Evaluate(T) (bool, error)
+}
+
+type FilePermissionExpected struct {
```

Comment:
I'm a little iffy on having these two implementations inside the same file, but they're short enough that it doesn't really matter.

Reply:
I'm indifferent right now. We can revisit it after it gets messier